C++2b / C++23 / C17 - Clang macOS/iOS - MSVC VS2022 - VS

danoli3 opened the pull request:
## C++ 23 / C17
- Clang macOS/iOS 
- MSVC VS2022 - VS 

This PR Updating Templates. 

Apothecary already building bleeding libraries with C++23 standard / C++2b https://github.com/openframeworks/apothecary/pull/390

danoli3 commented:
Not sure why C++23 is not accepted in VS2022 with latest. It might be written with c++2b instead of c++23 

https://learn.microsoft.com/en-us/cpp/overview/what-s-new-for-visual-cpp-in-visual-studio?view=msvc-170
As its meant to be included in 2023 .. <img width="816" alt="Screenshot 2024-07-17 at 8 44 27 PM" src="https://github.com/user-attachments/assets/94ef3281-5df5-4ee3-add2-8632c8b6b7c5">



https://learn.microsoft.com/en-us/cpp/overview/install-c17-support?view=msvc-170&source=recommendations

danoli3 commented:
```
Severity	Code	Description	Project	File	Line	Suppression State	Details
Error	C2280	'std::basic_ostream<char,std::char_traits<char>> &std::operator <<<std::char_traits<char>>(std::basic_ostream<char,std::char_traits<char>> &,const wchar_t *)': attempting to reference a deleted function	openframeworksLib	\\mac\Home\SOURCE\openFrameworksmacOS\libs\openFrameworks\utils\ofLog.h	509		
```

fixing now with:
on latest with c++latest used (23) 
also set c17

danoli3 commented:
MSVC some issues in ofLog() 
``
template <class T>
ofLog & operator<<(const T & value) {
	message << value << getPadding();
	return *this;
}
```

```
Severity	Code	Description	Project	File	Line	Suppression State	Details
Error	MSB4025	The project file could not be loaded. Could not find file 'C:\srx\openFrameworks\examples\templates\emptyExample\emptyExample.vcxproj'.	emptyExample	C:\srx\openFrameworks\examples\templates\emptyExample\emptyExample.vcxproj	1		
Error	C2280	'std::basic_ostream<char,std::char_traits<char>> &std::operator <<<std::char_traits<char>>(std::basic_ostream<char,std::char_traits<char>> &,const wchar_t *)': attempting to reference a deleted function	openframeworksLib	C:\srx\openFrameworks\libs\openFrameworks\utils\ofLog.h	509		
Error	C2088	built-in operator '<<' cannot be applied to an operand of type 'std::stringstream'	openframeworksLib	C:\srx\openFrameworks\libs\openFrameworks\utils\ofLog.h	509		
```

<img width="927" alt="Screenshot 2024-07-18 at 1 48 00 AM" src="https://github.com/user-attachments/assets/78cac877-1bde-46ab-86cc-3ce7e6cd8ac6">

<img width="1067" alt="Screenshot 2024-07-18 at 1 47 13 AM" src="https://github.com/user-attachments/assets/e7b07218-2df2-49a5-8b6e-2973b0f5ae69">

2bbb commented:
https://learn.microsoft.com/en-us/cpp/overview/cpp-conformance-improvements-2019?view=msvc-170

> Traditionally, inserting a wchar_t into a std::ostream, and inserting char16_t or char32_t into a std::ostream or std::wostream, outputs its integral value. Inserting pointers to those character types outputs the pointer value. Programmers don't find either case intuitive. They often expect the standard library to transcode the character or null-terminated character string instead, and to output the result.
> The C++20 proposal P1423R3 adds deleted stream insertion operator overloads for these combinations of stream and character or character pointer types. Under /std:c++20 or /std:c++latest, the overloads make these insertions ill-formed, instead of behaving in what is likely an unintended manner. The compiler raises error C2280 when one is found. You can define the "escape hatch" macro _HAS_STREAM_INSERTION_OPERATORS_DELETED_IN_CXX20 to 1 to restore the old behavior. (The proposal also deletes stream insertion operators for char8_t. Our standard library implemented similar overloads when we added char8_t support, so the "wrong" behavior has never been available for char8_t.)

danoli3 commented:
_HAS_STREAM_INSERTION_OPERATORS_DELETED_IN_CXX20 macro fixes VS2022 c++latest for the moment with ofLog issue for now. Might be a good idea to get this in as well

danoli3 commented:
- Fixed Bash ci bleeding project generator commands for Windows.
- Disabled Linux64 -2204 (now just 2404)
# Updated VS2022 project files (template for PG / Core OF Project) 
- Updated  support for Unicode (C/C++), including paths for vs project linking 
- Updated for C++23 / C++ Modules 
- Updated for C17 (latest C standard) - fixes C issues

danoli3 commented on addons/ofxKinect/libs/libfreenect/src/loader.c:
```diff
@@ -169,7 +169,7 @@ FN_INTERNAL int upload_firmware(fnusb_dev* dev, char * filename) {
 					continue;
 				int homelen = strlen(home);
-				char* dotfolder = "/.libfreenect";
+				char dotfolder[14] = "/.libfreenect";
```
fixes C17 issues  with this addon - as char* became const in new standard. this fixes it

danoli3 commented on .github/workflows/of.yml:
```diff
@@ -118,15 +118,16 @@ jobs:
+      - name: Build DEBUG and RELEASE allAddonsExample
+        working-directory: ${{env.GITHUB_WORKSPACE}}
+        run: |
+          msbuild examples/templates/allAddonsExample/allAddonsExample.vcxproj /p:configuration=debug /p:platform=${{ matrix.platform }} /p:PlatformToolset=v143
```
added release and addon tests

danoli3 commented on addons/ofxKinect/libs/libfreenect/src/usb_libusb10.c:
```diff
@@ -585,7 +585,8 @@ FN_INTERNAL int fnusb_open_subdevices(freenect_device *dev, int index)
 			else
 			{
-				res = upload_firmware(&dev->usb_audio, "audios.bin");
+				char audiosBin[11] = "audios.bin";
```
fixes C17 issues

danoli3 commented on libs/openFrameworks/app/ofAppEGLWindow.cpp:
```diff
@@ -409,10 +409,10 @@ void ofAppEGLWindow::setup(const ofAppEGLWindowSettings & _settings) {
 	// bool bIsX11Available = (pDisplay != NULL);
 
-	bool bIsX11Available = getenv("DISPLAY") != NULL;
+	bool bIsX11Available = ofGetEnv("DISPLAY") != NULL;
```
use oF wrapper with exception handler

danoli3 commented on libs/openFrameworks/utils/ofUtils.cpp:
```diff
@@ -1132,7 +1132,7 @@ std::string ofGetEnv(const std::string & var, const std::string defaultValue) {
 	}
 #else
-	auto value = getenv(var.c_str());
+	auto value = std::getenv(var.c_str());
```
explicit

danoli3 commented on scripts/ci/vs/install_bleeding.sh:
```diff
@@ -15,25 +14,39 @@ cd projectGenerator
-# unzip projectGenerator-vs-gui.zip 2> /dev/null
-# rm projectGenerator-vs-gui.zip
+OFW_ROOT=$(cygpath -w "$PWD")
+PG_OFX_PATH="${OF_ROOT}\\projectGenerator\\projectGenerator.exe"
```
this fixed the Ci PG for MSBUILD paths

danoli3 commented on scripts/ci/vs/install_bleeding.sh:
```diff
@@ -15,25 +14,39 @@ cd projectGenerator
+for i in "${!PROJECTS[@]}"; do
+    PROJECT=${PROJECTS[i]}
+    ADDON=${ADDONS[i]}
+    OPTIONS="-o\"${OF_ROOT}\" -v -a\"$ADDON\" -p\"vs\"  -t\"\" \"${OF_ROOT}\\${PROJECT}\""
```
previously was missing -p and addons

danoli3 reviewed (commented):
Building amazing on VS2022

danoli3 commented:
C++23 lets go

danoli3 commented:
This solves issues building / fixes projectGenerator download scripts. Fixes 2204 actions.
